### PR TITLE
feature/auto logout pg b2c

### PR DESCRIPTION
**Pull request by bbodine1**
> https://www.pivotaltracker.com/n/projects/2581014/stories/186417466

## Release Notes

- **New Features**
  - Added a search functionality to the app.

### CODE REVIEW

File: apps/provider-gateway/src/app/app.component.html [hunk 1]
```diff
@@ -17,7 +17,7 @@
       />
     </svg>
     <p><b>You are not authorized to view this page.</b></p>
-    <p><button (click)="login()">Login</button></p>
+    <p><button (click)="loginRedirect()">Login</button></p>
```
Comment (bbodine1) [inline apps/provider-gateway/src/app/app.component.html:20]: Just a rename of the login function to reflect msal standards

File: apps/provider-gateway/src/main.ts [hunk 2]
```diff
@@ -8,9 +8,6 @@ if (environment.production) {
   enableProdMode();
 }
 
-const gatewayApiUrl = environment.NX_GATEWAY_API || '';
```
Comment (bbodine1) [inline apps/provider-gateway/src/main.ts:11]: Removed unused variables

File: apps/b2c/src/app/last-active/last-active.component.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import { Observable } from 'rxjs';
+
+// services
+import { LastActiveService } from '../services/last-active.service';
+import { CommonModule } from '@angular/common';
+import { MomentModule } from 'ngx-moment';
+
+@Component({
+  // eslint-disable-next-line @angular-eslint/component-selector
+  selector: 'last-active',
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/last-active/last-active.component.ts:11]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/app/messages/messages.component.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import { Component } from '@angular/core';
+import { MessageService } from '../services/message.service';
+import { MatButtonModule } from '@angular/material/button';
+import { MatCardModule } from '@angular/material/card';
+import { CommonModule } from '@angular/common';
+
+@Component({
+  // eslint-disable-next-line @angular-eslint/component-selector
+  selector: 'messages',
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/messages/messages.component.ts:9]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/app/services/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  loginPopup() {
+    if (this.msalGuardConfig.authRequest) {
+      this.msalService
+        .loginPopup({ ...this.msalGuardConfig.authRequest } as PopupRequest)
+        .pipe(
+          filter((msg: AuthenticationResult) => {
+            console.log(msg);
+            return msg.account !== undefined;
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/services/auth.service.ts:63]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/app/services/auth.service.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+          })
+        );
+    } else {
+      this.msalService.loginPopup();
+    }
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/services/auth.service.ts:68]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/app/services/auth.service.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  public logout(popup?: boolean) {
+    if (popup) {
+      this.msalService.logoutPopup({
+        mainWindowRedirectUri: '/',
+      });
+    } else {
+      this.msalService.logoutRedirect();
+    }
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/services/auth.service.ts:78]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/app/services/auth.service.ts [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  private getLoggedInFromLocalStorage(): boolean | null {
+    const valueFromStorage = localStorage.getItem(this.lsLoggedInKey);
+    if (!valueFromStorage) {
+      return null;
+    }
+
+    return !!valueFromStorage;
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/services/auth.service.ts:87]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/app/services/last-active.service.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  map,
+} from 'rxjs';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class LastActiveService {
+  private recordTimeoutMs = 500;
+  public lsLastActiveKey = '__lastActive';
+  private events: string[] = ['keydown', 'click', 'wheel', 'mousemove'];
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/services/last-active.service.ts:19]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/app/services/last-active.service.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    fromEvent<StorageEvent>(window, 'storage')
+      .pipe(
+        filter(
+          (event) =>
+            event.storageArea === localStorage &&
+            event.key === this.lsLastActiveKey &&
+            !!event.newValue
+        ),
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        map((event) => new Date(event.newValue!))
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/services/last-active.service.ts:47]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/app/services/auth.service.ts [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+      .subscribe(() => {
+        this.log('tick');
+        const currentDate = moment(new Date());
+        const lastActiveDate = this.lastActiveService.getLastActiveDate();
+        if (
+          moment.duration(currentDate.diff(lastActiveDate)).asSeconds() >
+          this.inactivityLogoutTimeoutS
+        ) {
+          this.logout();
+          // window.location.reload();
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/services/auth.service.ts:105]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/app/services/auth.service.ts [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        }
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/services/auth.service.ts:106]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/app/app.component.ts [hunk 13]
```diff
@@ -140,51 +141,41 @@
...
-      this.authService.logoutPopup({
-        mainWindowRedirectUri: '/',
-      });
-    } else {
-      this.authService.logoutRedirect();
-    }
+    this.ourAuthService.logout(popup);
   }
 
   ngOnDestroy(): void {
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/app.component.ts:181]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/environments/environment.development.ts [hunk 14]
```diff
@@ -11,4 +11,7 @@ export const environment = {
     scopes: ['user.read'],
     uri: 'https://graph.microsoft.com/v1.0/me',
   },
+  apiUrl: 'https://localhost:7103',
+  // appInactiveTimeout: 900, // 15 minutes
+  appInactiveTimeout: 10, // 15 minutes
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/environments/environment.development.ts:16]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/provider-gateway/src/app/services/last-active.service.ts [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+  throttleTime,
+  filter,
+  map,
+} from 'rxjs';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class LastActiveService {
+  private recordTimeoutMs = 500;
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/provider-gateway/src/app/services/last-active.service.ts:17]: This is the correct value

File: apps/provider-gateway/src/app/services/last-active.service.ts [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    fromEvent<StorageEvent>(window, 'storage')
+      .pipe(
+        filter(
+          (event) =>
+            event.storageArea === localStorage &&
+            event.key === this.lsLastActiveKey &&
+            !!event.newValue
+        ),
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        map((event) => new Date(event.newValue!))
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/provider-gateway/src/app/services/last-active.service.ts:47]: the pipe filter before the map using this function checks to make sure there is an actual value (`!!event.newValue`). This insures that the value passed to the map is never null but always truthy. This is ok.

File: apps/provider-gateway/src/app/services/auth.service.ts [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+    }
+  }
+
+  private getLoggedInFromLocalStorage(): boolean | null {
+    const valueFromStorage = localStorage.getItem(this.lsLoggedInKey);
+    if (!valueFromStorage) {
+      return null;
+    }
+
+    return !!valueFromStorage;
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/provider-gateway/src/app/services/auth.service.ts:85]: There is no data stored in local storage that could affect security of the application. This is part of an oauth JWT authentication pattern. This is ok.

File: apps/provider-gateway/src/app/services/auth.service.ts [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  loginPopup() {
+    if (this.msalGuardConfig.authRequest) {
+      this.msalService
+        .loginPopup({ ...this.msalGuardConfig.authRequest } as PopupRequest)
+        .pipe(
+          filter((msg: AuthenticationResult) => {
+            console.log(msg);
+            return msg.account !== undefined;
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/provider-gateway/src/app/services/auth.service.ts]: Removed

File: apps/provider-gateway/src/app/services/auth.service.ts [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+        filter(
+          (event) =>
+            event.storageArea === localStorage &&
+            event.key === this.lsLoggedInKey
+        ),
+        map((event) => !!event.newValue)
+      )
+      .subscribe((loggedIn) => {
+        this._loggedIn.next(loggedIn);
+      });
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/provider-gateway/src/app/services/auth.service.ts:113]: This subscription will never be unsubscribed from.

File: apps/provider-gateway/src/app/services/auth.service.ts [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+    const valueFromStorage = localStorage.getItem(this.lsLoggedInKey);
+    if (!valueFromStorage) {
+      return null;
+    }
+
+    return !!valueFromStorage;
+  }
+
+  private inactivityLogoutTimeoutS = environment.appInactiveTimeout;
+  private timerTickMs = 1000;
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/provider-gateway/src/app/services/auth.service.ts:84]: There is no need to change this ticker. It is part of the clock function that runs every second to check if a user is logged in.

File: apps/provider-gateway/src/app/services/auth.service.ts [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+  public login() {
+    if (this.msalGuardConfig.authRequest) {
+      this.msalService.loginRedirect({
+        ...this.msalGuardConfig.authRequest,
+        account: this.msalService.instance.getActiveAccount(),
+        loginHint: this.msalService.instance.getActiveAccount()?.username,
+      } as RedirectRequest);
+    } else {
+      this.msalService.loginRedirect();
+    }
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/provider-gateway/src/app/services/auth.service.ts]: added error check

File: apps/provider-gateway/src/app/services/auth.service.ts [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+      .subscribe(() => {
+        // console.log('tick'); // for debugging
+        const currentDate = moment(new Date());
+        const lastActiveDate = this.lastActiveService.getLastActiveDate();
+        if (
+          moment.duration(currentDate.diff(lastActiveDate)).asSeconds() >
+          this.inactivityLogoutTimeoutS
+        ) {
+          this.logout();
+        }
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/provider-gateway/src/app/services/auth.service.ts:104]: "moment" is the package of choice for now.

File: apps/b2c/project.json [hunk 23]
```diff
@@ -8,25 +8,17 @@
...
-          "apps/b2c/src/assets"
-        ],
-        "styles": [
-          "apps/b2c/src/styles.scss"
-        ],
-        "scripts": []
+        "assets": ["apps/b2c/src/favicon.ico", "apps/b2c/src/assets"],
+        "styles": ["apps/b2c/src/styles.scss"],
+        "scripts": [],
+        "allowedCommonJsDependencies": ["moment"]
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/project.json:21]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/provider-gateway/src/environments/environment.prod.ts [hunk 24]
```diff
@@ -1,9 +1,33 @@
...
+const gatewayApiUrl = 'https://provider-gateway-api.dbh.dc.gov';
+const portalApiUrl = 'https://ite-api.dbh.dc.gov';
+const authority = `https://${subdomain}.b2clogin.com/${subdomain}.onmicrosoft.com/b2c_1_sign_in_1`;
+const authorityDomain = `${subdomain}.b2clogin.com`;
+// const appInactiveTimeout = 900; // 15 minutes
+const appInactiveTimeout = 10; // 10 seconds
+
 export const environment = {
-  production: true,
+  production: false,
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/provider-gateway/src/environments/environment.prod.ts:11]: This is part of the demo app. No updates. It is working and will be removed soon.

File: apps/b2c/src/environments/environment.ts [hunk 25]
```diff
@@ -15,6 +15,8 @@ export const environment = {
     scopes: [],
     uri: '',
   },
+  apiUrl: 'https://localhost:7103',
```
Comment (kristinmerbach) [inline apps/b2c/src/environments/environment.ts:18]: remove apiUrl

File: apps/provider-gateway/src/environments/environment.ts [hunk 26]
```diff
@@ -1,15 +1,39 @@
...
+      clientId: clientId,
+      authority: authority,
+      authorityDomain: authorityDomain,
+    },
+  },
+  // apiConfig: {
+  //   scopes: ['user.read'],
+  //   uri: 'https://graph.microsoft.com/v1.0/me',
+  // },
+  // apiUrl: 'https://localhost:7103',
```
Comment (kristinmerbach) [inline apps/provider-gateway/src/environments/environment.ts:35]: remove api url reference entirely

File: apps/provider-gateway/src/environments/environment.ts [hunk 27]
```diff
@@ -1,15 +1,39 @@
...
+  subdomain: subdomain,
+
+  msalConfig: {
+    auth: {
+      clientId: clientId,
+      authority: authority,
+      authorityDomain: authorityDomain,
+    },
+  },
+  // apiConfig: {
```
Comment (kristinmerbach) [inline apps/provider-gateway/src/environments/environment.ts:31]: uncomment apiConfig and set to suggestion:
```suggestion
  apiConfig: {
    scopes: ['provider.read'],
    uri: `https://${subdomain}.onmicrosoft.com/provider-api/`,
  },
```

File: apps/provider-gateway/src/environments/environment.prod.ts [hunk 28]
```diff
@@ -1,9 +1,33 @@
...
+  subdomain: subdomain,
+
+  msalConfig: {
+    auth: {
+      clientId: clientId,
+      authority: authority,
+      authorityDomain: authorityDomain,
+    },
+  },
+  // apiConfig: {
```
Comment (kristinmerbach) [inline apps/provider-gateway/src/environments/environment.prod.ts:27]:
```suggestion
 apiConfig: {
    scopes: ['provider.read'],
    uri: `https://${subdomain}.onmicrosoft.com/provider-api/`,
  },
```

File: apps/provider-gateway/src/environments/environment.prod.ts [hunk 29]
```diff
@@ -1,9 +1,33 @@
...
+  //   scopes: ['user.read'],
+  //   uri: 'https://graph.microsoft.com/v1.0/me',
+  // },
+  // apiUrl: 'https://localhost:7103',
```
Comment (kristinmerbach) [inline apps/provider-gateway/src/environments/environment.prod.ts:31]: remove apiUrl variable

File: apps/b2c/src/environments/environment.development.ts [hunk 30]
```diff
@@ -11,4 +11,7 @@ export const environment = {
     scopes: ['user.read'],
```
Comment (kristinmerbach) [inline apps/b2c/src/environments/environment.development.ts:11]:
```suggestion
    apiConfig: {
    scopes: ['provider.read'],
    uri: `https://${subdomain}.onmicrosoft.com/provider-api/`,
  },
```

File: apps/b2c/src/environments/environment.development.ts [hunk 31]
```diff
@@ -11,4 +11,7 @@ export const environment = {
...
     uri: 'https://graph.microsoft.com/v1.0/me',
   },
+  apiUrl: 'https://localhost:7103',
```
Comment (kristinmerbach) [inline apps/b2c/src/environments/environment.development.ts:14]: remove apiUrl reference

File: apps/provider-gateway/src/app/components/login-failed.component.ts [hunk 32]
```diff
@@ -0,0 +1,15 @@
+import { Component } from '@angular/core';
+import { CommonModule } from '@angular/common';
+import { RouterLink } from '@angular/router';
+
+@Component({
+  selector: 'dbh-login-failed',
+  standalone: true,
+  imports: [CommonModule, RouterLink],
+  template: `
+    <p>login-failed works!</p>
```
Comment (kristinmerbach) [inline apps/provider-gateway/src/app/components/login-failed.component.ts:10]: remove this p and text

**Review by kristinmerbach — CHANGES_REQUESTED** [review]
This is amazing work. A few comments but nothing major.

File: apps/b2c/src/app/app.component.ts [hunk 33]
```diff
@@ -140,51 +141,41 @@ export class AppComponent implements OnInit, OnDestroy {
...
     }
-  }
 
-  loginRedirect() {
-    if (this.msalGuardConfig.authRequest) {
-      this.authService.loginRedirect({
-        ...this.msalGuardConfig.authRequest,
-      } as RedirectRequest);
+    if (activeAccount) {
+      console.log('We are logged in');
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/app/app.component.ts:155]: image
Anything that is in the b2c app we will remove soon. I don't think we should update it at all.

File: apps/b2c/src/main.ts [hunk 34]
```diff
@@ -51,7 +57,7 @@ export function MSALInstanceFactory(): IPublicClientApplication {
       clientId: environment.msalConfig.auth.clientId,
       authority: environment.msalConfig.auth.authority,
       redirectUri: '/',
-      postLogoutRedirectUri: '/',
+      postLogoutRedirectUri: 'https://localhost:4202/',
```
[the remark this answers is not in the document]
Comment (bbodine1) [reply apps/b2c/src/main.ts:60]: This is b2c demo app code. We'll remove it once we get provider-gateway and probably portal auto-logouting

File: apps/b2c/src/environments/environment.development.ts (hunk 31 above)
Comment (bbodine1) [reply apps/b2c/src/environments/environment.development.ts:17]: b2c demo app code

File: apps/b2c/src/environments/environment.ts (hunk 25 above)
Comment (bbodine1) [reply apps/b2c/src/environments/environment.ts:18]: b2c demo app code

**Review by kristinmerbach — APPROVED** [review]

**Pull request merged**